File: src/main/java/br/com/leonardoferreira/jirareport/config/security/JiraAuthenticationProvider.java
package br.com.leonardoferreira.jirareport.config.security;

import br.com.leonardoferreira.jirareport.domain.form.LoginForm;
import br.com.leonardoferreira.jirareport.domain.vo.Account;
import br.com.leonardoferreira.jirareport.service.AuthService;
import br.com.leonardoferreira.jirareport.service.TokenService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;

@Slf4j
@Component
public class JiraAuthenticationProvider implements AuthenticationProvider {

    @Autowired
    private AuthService authService;

    @Autowired
    private TokenService tokenService;

    @Override
    public Authentication authenticate(final Authentication auth) {
        try {
            Account account = authService.login(new LoginForm(auth.getName(), auth.getCredentials().toString()));
            if (account == null) {
                throw new BadCredentialsException("External system authentication failed");
            }

            String token = tokenService.encode(account);

            UsernamePasswordAuthenticationToken user = new UsernamePasswordAuthenticationToken(
                    account, account.getPassword(), account.getAuthorities());
            user.setDetails(token);

            return user;
        } catch (Exception e) {
            log.error("Method=authenticate, E={}", e.getMessage(), e);
            throw new BadCredentialsException(e.getMessage(), e);
        }
    }

    @Override
    public boolean supports(final Class<?> auth) {
        return auth.equals(UsernamePasswordAuthenticationToken.class);
    }
}

File: application/src/main/kotlin/br/com/jiratorio/security/JiraAuthenticationProvider.kt
package br.com.jiratorio.security

import br.com.jiratorio.domain.request.LoginRequest
import br.com.jiratorio.service.AuthService
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component

@Component
class JiraAuthenticationProvider(
    private val authService: AuthService,
) : AuthenticationProvider {

    private val log = LoggerFactory.getLogger(javaClass)

    override fun authenticate(auth: Authentication): Authentication {
        try {
            val account = authService.login(LoginRequest(auth.name, auth.credentials.toString()))
            val token = authService.encode(account)

            return UsernamePasswordAuthenticationToken(account, account.password, account.authorities)
                .also { it.details = token }
        } catch (e: Exception) {
            log.error("Method=authenticate, E={}", e.message, e)
            throw BadCredentialsException(e.message, e)
        }
    }

    override fun supports(auth: Class<*>): Boolean =
        auth == UsernamePasswordAuthenticationToken::class.java
}
